### ISSUE TITLE
The library can't handle large lists

### ISSUE THREAD

**Issue by sveta-slepner** [TanStack/virtual#565]
### Describe the bug

When trying to create a list with a large count of rows, one of two things happens:

A. If the count is over an arbitrary number, capped by each browser individually, the scroll won't go over that row.
For instance, for Chrome that number is 479348, and you can't scroll to the remaining items if there are any.

B. If the number is too large (for instance 150 mil), the tab will simply crash.
I believe this is caused by the `getMeasurements` function from the `virtual-core` library, as it tries to create an array of the count size.

### Your minimal, reproducible example

https://codesandbox.io/p/sandbox/silent-wildflower-594ycm

### Steps to reproduce

1. In the sandbox click on one of the example buttons that will set a larger number, or edit the count property manually.

### Expected behavior

A. I expect the scroll to allow the user to reach the exact number specified in the count.
B. I expect the browser not to crash .

### How often does this bug happen?

Every time

### Platform

- Any but specifically I'm using:
Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/114.0.0.0 Safari/537.36

### tanstack-virtual version

v3.0.0-beta.54

### Terms & Code of Conduct

- [X] I agree to follow this project's Code of Conduct
- [X] I understand that if my bug cannot be reliable reproduced in a debuggable environment, it will probably not be fixed and this issue may even be closed.

**Comment by ddolcimascolo** [comment]
Hi all,

@tannerlinsley I just started using the library (the React binding, specifically) to implement an infinite list component and it works pretty well. I really like the approach of providing the basic feature of virtualizing a large list and letting us implement the rendering, fetching, etc.

I experienced this issue, and enabling `debug: true` show that all the time is spent is `getMeasurements` as @sveta-slepner mentioned, I have timings around 100ms in this method, and the scrolling is laggy at best :) I worked around this problem by providing a smaller `count` to the virtualizer, increasing it as more pages are fetched. Scrolling is fast even with thousands of items loaded.

Anything I can do to help fix that?

As a side note, when are you planning to release a stable v3?

Thanks,
David

**Comment by tannerlinsley** [comment]
I'll take a look at this asap.

**Comment by knownasilya** [comment]
Cross posting from discord.

I think the virtualizer holds on to dom nodes. Example app: https://stackblitz.com/edit/sb1-41eywcye?file=index.html
I changed pages from home to project a few times and this is the result, I'd expect the number of nodes to go down. In my own app I see this but it keeps growing even after 1gb of memory.

<img width="1110" alt="Image" src="https://github.com/user-attachments/assets/f7a8a5d5-f52d-439a-99e0-adda53da2eab" />

**Comment by piecyk** [comment]
> I think the virtualizer holds on to dom nodes.

that is true, we store it via measureElement https://github.com/TanStack/virtual/blob/main/packages/virtual-core/src/index.ts#L753

and this part clears it via `isConnected` https://github.com/TanStack/virtual/blob/main/packages/virtual-core/src/index.ts#L793-L801

**Comment by knownasilya** [comment]
@piecyk hum, I don't see any cleanup on unmount though. Am I missing it or does it not happen?

**Comment by piecyk** [comment]
@knownasilya yeah there is any, the idea was that they will be cleared by measureElement, but let's also clear it in cleanup.

> This line https://github.com/TanStack/virtual/blob/main/packages/virtual-core/src/index.ts#L101 could probably be observer.disconnect()

could be, but there should not be any difference because we always observe one rect element

**Comment by piecyk** [comment]
> I think the virtualizer holds on to dom nodes. Example app: https://stackblitz.com/edit/sb1-41eywcye?file=index.html

@knownasilya example not found?

**Comment by knownasilya** [comment]
@piecyk I made it public, should be good now

**Comment by piecyk** [comment]
> [@piecyk](https://github.com/piecyk) I made it public, should be good now

Thanks, from quick look noticed that there is pretty old version of @tanstack/react-virtual, please update to latest, also clicked around and didn't noticed memory leaking.

**Comment by knownasilya** [comment]
I updated the version and now I can't reproduce the leak, so maybe we're good? Bad news for me, probably means there is another issue in my app.<img width="836" alt="Image" src="https://github.com/user-attachments/assets/8ecd6619-3ec9-47d5-b126-188996654c45" />

**Comment by knownasilya** [comment]
Might be a react 17 vs 18.. https://github.com/facebook/react/pull/17666

**Comment by geferon** [comment]
> I updated the version and now I can't reproduce the leak, so maybe we're good? Bad news for me, probably means there is another issue in my app.<img alt="Image" width="836" src="https://private-user-images.githubusercontent.com/34726/398421695-8ecd6619-3ec9-47d5-b126-188996654c45.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzU1ODY1NzgsIm5iZiI6MTczNTU4NjI3OCwicGF0aCI6Ii8zNDcyNi8zOTg0MjE2OTUtOGVjZDY2MTktM2VjOS00N2Q1LWIxMjYtMTg4OTk2NjU0YzQ1LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDEyMzAlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQxMjMwVDE5MTc1OFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTkzYjQzZWE5ZWZkNTNhNDY0MjRmOTZjZmVmMjliZTVmZmQyZjliYTg1NTBmN2YzZjI1NTEwNjMzMGI5M2NiNmMmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.UY5TR5MKhbiBEWA9zJRJrjFF_agjcyCVujUyTDBI9X0">

I'm having this exact same issue in Vue as well, so I'm not so sure if it's a framework specific issue?

**Comment by geferon** [comment]
![Image](https://github.com/user-attachments/assets/bbe6a646-d85f-4ea6-bc10-a9bd0b07411f)
The spikes going down is me changing pages, removing and re-adding the component, and this is in Vue.

**Comment by Enable-Lukas** [comment]
Hi, I also noticed the issue with scrolling.
I am using Svelte as a framework. I'm not very experienced when it comes to reading memory usage, but from what I can tell, everything looks fine on my side. I don't think the issue is related to a memory leak.
### 3 Page Loads

![Image](https://github.com/user-attachments/assets/e212f2af-6dd6-46ee-b66f-1e1153051962)

What also supports this is the fact that the same number of rows is displayed every time I reload the page. If it were a memory leak, I would expect the number of rows to decrease with each reload.

I hope this information is somewhat helpful. As I mentioned, I’m still new to all of this, so I might be completely wrong.

**Comment by joekaiser** [comment]
I work for a large enterprise company and ran into this problem after upgrading to the latest version (3.13.6). We were seeing multiple GB of memory usage. We "solved" this by calling `measureElement(null)`. This will force a cleanup of the elementsCache. [see commit here](https://github.com/TanStack/virtual/pull/801/files#diff-85d1b85f3d4e501779ca41e51cd38cf7f61a2b7ffddaf958cd7e79c2787c36f1R775)

We decided to call this in the `onChange` event of the virtualizer itself. We just call it anytime `sync` (i.e. scrolling) is false.         
```
onChange: (virtualizer, sync) => {
    if (!sync) {
        virtualizer.measureElement(null);
    }
},
```

Quite the hacky patch, but it stopped the bleeding 🤷‍♂️

**Comment by piecyk** [comment]
@joekaiser thanks for reporting this, shouldn't ref callbacks be triggered with null when elements are unmounted? I would have expected that to eventually lead to cleanup via measureElement(null), especially in React 16. Is there a reason this doesn’t happen, or is the virtualizer holding onto refs longer than expected?

**Comment by gavinwahl** [comment]
This is happening for me on react 18.2 and react-virtual 3.13.10
